src/snowflake/snowpark/_internal/compiler/query_generator.py

```diff
@@ -169,6 +169,7 @@ def do_resolve_with_resolved_children(
                 iceberg_config=logical_plan.iceberg_config,
                 table_exists=logical_plan.table_exists,
             )
+            resolved_plan.referenced_ctes = resolved_child.referenced_ctes
```

Comment:
it seems we missed the cte reference propagation in the plan builder for some case, we should double check the PlanBuilder code in snowflake_plan, instead of doing the fix here

Reply:
After some thoughts, let's update the referenced_ctes definition to the cte referenced in the plan tree, and let's propagate  the ctes  reference for all nodes, in other words, let's deprecate this parameter https://github.com/snowflakedb/snowpark-python/blob/main/src/snowflake/snowpark/_internal/analyzer/snowflake_plan.py#L538, and make sure the referenced ctes is propagated correctly for all plan builder, please double check all places that directly create SnowflakePlan to make sure the referenced_ctes is propagated correctly.

Reply:
For the correction of code generation, we can use the way you have now, but please make sure we comment it out clearly about why things is done in such way

Reply:
If the cte reference is propagated correctly during plan builder, you shouldn't need to re-propagate here.